File: app/src/main/java/com/example/back2me/AddEditItemActivity.java
package com.example.back2me;

import android.Manifest;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import com.bumptech.glide.Glide;
import com.example.back2me.databinding.ActivityAddEditItemBinding;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;

public class AddEditItemActivity extends AppCompatActivity {

    private static final String TAG = "AddEditItemActivity";

    private ActivityAddEditItemBinding binding;
    private FirebaseAuth auth;
    private Handler mainHandler;

    private String selectedStatus = "lost";
    private Uri selectedImageUri = null;
    private Bitmap selectedBitmap = null;
    private boolean isUploading = false;

    // Camera launcher
    private final ActivityResultLauncher<Void> cameraLauncher =
            registerForActivityResult(new ActivityResultContracts.TakePicturePreview(), bitmap -> {
                if (bitmap != null) {
                    selectedBitmap = bitmap;
                    selectedImageUri = null;
                    displaySelectedImage(bitmap);
                    Log.d(TAG, "Camera image captured");
                }
            });

    // Gallery launcher
    private final ActivityResultLauncher<String> galleryLauncher =
            registerForActivityResult(new ActivityResultContracts.GetContent(), uri -> {
                if (uri != null) {
                    selectedImageUri = uri;
                    selectedBitmap = null;
                    displaySelectedImage(uri);
                    Log.d(TAG, "Gallery image selected: " + uri);
                }
            });

    // Camera permission launcher
    private final ActivityResultLauncher<String> cameraPermissionLauncher =
            registerForActivityResult(new ActivityResultContracts.RequestPermission(), isGranted -> {
                if (isGranted) {
                    cameraLauncher.launch(null);
                } else {
                    Toast.makeText(this, R.string.camera_permission_required, Toast.LENGTH_SHORT).show();
                }
            });

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityAddEditItemBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        auth = FirebaseAuth.getInstance();
        mainHandler = new Handler(Looper.getMainLooper());

        setupUI();
        setupClickListeners();
    }

    private void setupUI() {
        binding.chipLost.setChecked(true);
        selectedStatus = "lost";
        binding.cardImagePreview.setVisibility(View.GONE);
    }

    private void setupClickListeners() {
        binding.backButton.setOnClickListener(v -> finish());

        binding.chipLost.setOnClickListener(v -> {
            selectedStatus = "lost";
            binding.chipLost.setChecked(true);
            binding.chipFound.setChecked(false);
        });

        binding.chipFound.setOnClickListener(v -> {
            selectedStatus = "found";
            binding.chipFound.setChecked(true);
            binding.chipLost.setChecked(false);
        });

        binding.buttonAddPhoto.setOnClickListener(v -> showImagePickerDialog());
        binding.buttonRemovePhoto.setOnClickListener(v -> removeSelectedImage());
        binding.buttonSubmit.setOnClickListener(v -> validateAndSubmit());
    }

    private void showImagePickerDialog() {
        String[] options = {"Take Photo", "Choose from Gallery"};

        new AlertDialog.Builder(this)
                .setTitle("Add Photo")
                .setItems(options, (dialog, which) -> {
                    if (which == 0) {
                        checkCameraPermissionAndLaunch();
                    } else {
                        galleryLauncher.launch("image/*");
                    }
                })
                .show();
    }

    private void checkCameraPermissionAndLaunch() {
        if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                == PackageManager.PERMISSION_GRANTED) {
            cameraLauncher.launch(null);
        } else {
            cameraPermissionLauncher.launch(Manifest.permission.CAMERA);
        }
    }

    private void displaySelectedImage(Bitmap bitmap) {
        binding.cardImagePreview.setVisibility(View.VISIBLE);
        binding.imagePreview.setImageBitmap(bitmap);
        binding.buttonAddPhoto.setText(R.string.change_photo);
    }

    private void displaySelectedImage(Uri uri) {
        binding.cardImagePreview.setVisibility(View.VISIBLE);
        Glide.with(this)
                .load(uri)
                .centerCrop()
                .into(binding.imagePreview);
        binding.buttonAddPhoto.setText(R.string.change_photo);
    }

    private void removeSelectedImage() {
        selectedImageUri = null;
        selectedBitmap = null;
        binding.cardImagePreview.setVisibility(View.GONE);
        binding.imagePreview.setImageDrawable(null);
        binding.buttonAddPhoto.setText(R.string.add_photo);
    }

    private void validateAndSubmit() {
        String name = binding.editItemName.getText().toString().trim();
        String location = binding.editLocation.getText().toString().trim();
        String description = binding.editDescription.getText().toString().trim();

        if (name.isEmpty()) {
            binding.editItemName.setError(getString(R.string.error_item_name_required));
            binding.editItemName.requestFocus();
            return;
        }

        if (location.isEmpty()) {
            binding.editLocation.setError(getString(R.string.error_location_required));
            binding.editLocation.requestFocus();
            return;
        }

        if (isUploading) {
            Toast.makeText(this, R.string.please_wait, Toast.LENGTH_SHORT).show();
            return;
        }

        FirebaseUser currentUser = auth.getCurrentUser();
        if (currentUser == null) {
            Toast.makeText(this, R.string.please_login, Toast.LENGTH_SHORT).show();
            return;
        }

        setLoading(true);

        // Check if we have an image to upload
        if (selectedBitmap != null) {
            uploadBitmapToCloudinary(name, location, description, currentUser.getUid());
        } else if (selectedImageUri != null) {
            uploadUriToCloudinary(name, location, description, currentUser.getUid());
        } else {
            createItem(name, location, description, currentUser.getUid(), "");
        }
    }

    private void uploadBitmapToCloudinary(String name, String location, String description, String userId) {
        isUploading = true;
        Log.d(TAG, "Uploading bitmap to Cloudinary...");

        CloudinaryHelper.uploadBitmap(selectedBitmap, new CloudinaryHelper.UploadCallback() {
            @Override
            public void onSuccess(String imageUrl) {
                mainHandler.post(() -> {
                    isUploading = false;
                    Log.d(TAG, "Cloudinary upload success: " + imageUrl);
                    createItem(name, location, description, userId, imageUrl);
                });
            }

            @Override
            public void onError(String errorMessage) {
                mainHandler.post(() -> {
                    isUploading = false;
                    setLoading(false);
                    Log.e(TAG, "Cloudinary upload error: " + errorMessage);
                    Toast.makeText(AddEditItemActivity.this,
                            "Failed to upload image: " + errorMessage, Toast.LENGTH_LONG).show();
                });
            }
        });
    }

    private void uploadUriToCloudinary(String name, String location, String description, String userId) {
        isUploading = true;
        Log.d(TAG, "Uploading URI to Cloudinary...");

        CloudinaryHelper.uploadUri(this, selectedImageUri, new CloudinaryHelper.UploadCallback() {
            @Override
            public void onSuccess(String imageUrl) {
                mainHandler.post(() -> {
                    isUploading = false;
                    Log.d(TAG, "Cloudinary upload success: " + imageUrl);
                    createItem(name, location, description, userId, imageUrl);
                });
            }

            @Override
            public void onError(String errorMessage) {
                mainHandler.post(() -> {
                    isUploading = false;
                    setLoading(false);
                    Log.e(TAG, "Cloudinary upload error: " + errorMessage);
                    Toast.makeText(AddEditItemActivity.this,
                            "Failed to upload image: " + errorMessage, Toast.LENGTH_LONG).show();
                });
            }
        });
    }

    private void createItem(String name, String location, String description,
                            String userId, String imageUrl) {
        SimpleDateFormat isoFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US);
        isoFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
        String createdDate = isoFormat.format(new Date());

        Item item = new Item(
                "",
                name,
                location,
                description,
                selectedStatus,
                userId,
                createdDate,
                imageUrl
        );

        Log.d(TAG, "Creating item with imageUrl: " + (imageUrl.isEmpty() ? "(none)" : imageUrl));

        ItemRepository.createItem(item, new ItemRepository.CreateCallback() {
            @Override
            public void onSuccess(Item createdItem) {
                setLoading(false);
                Log.d(TAG, "Item created successfully: " + createdItem.getId());
                Toast.makeText(AddEditItemActivity.this,
                        R.string.item_posted_successfully, Toast.LENGTH_SHORT).show();
                finish();
            }

            @Override
            public void onError(Exception e) {
                setLoading(false);
                Log.e(TAG, "Error creating item", e);
                Toast.makeText(AddEditItemActivity.this,
                        getString(R.string.error_posting_item) + ": " + e.getMessage(),
                        Toast.LENGTH_LONG).show();
            }
        });
    }

    private void setLoading(boolean loading) {
        binding.progressBar.setVisibility(loading ? View.VISIBLE : View.GONE);
        binding.buttonSubmit.setEnabled(!loading);
        binding.buttonSubmit.setText(loading ? R.string.posting : R.string.post_item);
        binding.editItemName.setEnabled(!loading);
        binding.editLocation.setEnabled(!loading);
        binding.editDescription.setEnabled(!loading);
        binding.chipLost.setEnabled(!loading);
        binding.chipFound.setEnabled(!loading);
        binding.buttonAddPhoto.setEnabled(!loading);
    }
}

File: app/src/main/java/com/example/back2me/AddEditItemActivity.kt
package com.example.back2me

import android.Manifest
import android.app.DatePickerDialog
import android.app.TimePickerDialog
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.view.View
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import com.example.back2me.ItemRepository
import com.example.back2me.databinding.ActivityAddEditItemBinding
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.storage.FirebaseStorage
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import java.io.ByteArrayOutputStream
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Locale
import java.util.TimeZone
import java.util.UUID

class AddEditItemActivity : AppCompatActivity() {

    private lateinit var binding: ActivityAddEditItemBinding
    private var selectedDate: Calendar = Calendar.getInstance()
    private var selectedImageUri: Uri? = null
    private var uploadedImageUrl: String = ""

    // Camera launcher
    private val cameraLauncher = registerForActivityResult(ActivityResultContracts.TakePicturePreview()) { bitmap ->
        if (bitmap != null) {
            displayImage(bitmap)
        }
    }

    // Gallery launcher
    private val galleryLauncher = registerForActivityResult(ActivityResultContracts.GetContent()) { uri ->
        if (uri != null) {
            selectedImageUri = uri
            binding.imagePreview.setImageURI(uri)
            binding.imagePreview.visibility = View.VISIBLE
            binding.uploadPrompt.visibility = View.GONE
            binding.buttonRemovePhoto.visibility = View.VISIBLE
        }
    }

    // Permission launcher
    private val cameraPermissionLauncher = registerForActivityResult(ActivityResultContracts.RequestPermission()) { isGranted ->
        if (isGranted) {
            openCamera()
        } else {
            Toast.makeText(this, "Camera permission denied", Toast.LENGTH_SHORT).show()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAddEditItemBinding.inflate(layoutInflater)
        setContentView(binding.root)

        setupClickListeners()
        setDefaultDateTime()
    }

    private fun setupClickListeners() {
        binding.backButton.setOnClickListener { finish() }
        binding.buttonCancel.setOnClickListener { finish() }
        binding.buttonSaveItem.setOnClickListener { postItem() }

        // Date & Time
        binding.inputDate.setOnClickListener { showDatePicker() }
        binding.inputTime.setOnClickListener { showTimePicker() }

        // Photo actions - Show dialog when clicking photo card
        binding.photoCard.setOnClickListener { showPhotoOptionsDialog() }
        binding.buttonRemovePhoto.setOnClickListener { removePhoto() }
    }

    private fun showPhotoOptionsDialog() {
        val options = arrayOf("📷 Take Photo", "🖼️ Choose from Gallery")

        AlertDialog.Builder(this)
            .setTitle("Add Photo")
            .setItems(options) { _, which ->
                when (which) {
                    0 -> checkCameraPermissionAndOpen() // Camera
                    1 -> galleryLauncher.launch("image/*") // Gallery
                }
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun setDefaultDateTime() {
        val dateFormat = SimpleDateFormat("MMM dd, yyyy", Locale.getDefault())
        val timeFormat = SimpleDateFormat("hh:mm a", Locale.getDefault())

        binding.inputDate.setText(dateFormat.format(selectedDate.time))
        binding.inputTime.setText(timeFormat.format(selectedDate.time))
    }

    private fun showDatePicker() {
        val year = selectedDate.get(Calendar.YEAR)
        val month = selectedDate.get(Calendar.MONTH)
        val day = selectedDate.get(Calendar.DAY_OF_MONTH)

        DatePickerDialog(this, { _, selectedYear, selectedMonth, selectedDay ->
            selectedDate.set(Calendar.YEAR, selectedYear)
            selectedDate.set(Calendar.MONTH, selectedMonth)
            selectedDate.set(Calendar.DAY_OF_MONTH, selectedDay)

            val dateFormat = SimpleDateFormat("MMM dd, yyyy", Locale.getDefault())
            binding.inputDate.setText(dateFormat.format(selectedDate.time))
        }, year, month, day).show()
    }

    private fun showTimePicker() {
        val hour = selectedDate.get(Calendar.HOUR_OF_DAY)
        val minute = selectedDate.get(Calendar.MINUTE)

        TimePickerDialog(this, { _, selectedHour, selectedMinute ->
            selectedDate.set(Calendar.HOUR_OF_DAY, selectedHour)
            selectedDate.set(Calendar.MINUTE, selectedMinute)

            val timeFormat = SimpleDateFormat("hh:mm a", Locale.getDefault())
            binding.inputTime.setText(timeFormat.format(selectedDate.time))
        }, hour, minute, false).show()
    }

    private fun checkCameraPermissionAndOpen() {
        when {
            ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA) == PackageManager.PERMISSION_GRANTED -> {
                openCamera()
            }
            else -> {
                cameraPermissionLauncher.launch(Manifest.permission.CAMERA)
            }
        }
    }

    private fun openCamera() {
        cameraLauncher.launch(null)
    }

    private fun displayImage(bitmap: Bitmap) {
        binding.imagePreview.setImageBitmap(bitmap)
        binding.imagePreview.visibility = View.VISIBLE
        binding.uploadPrompt.visibility = View.GONE
        binding.buttonRemovePhoto.visibility = View.VISIBLE

        // Convert bitmap to URI for upload
        selectedImageUri = getImageUriFromBitmap(bitmap)
    }

    private fun getImageUriFromBitmap(bitmap: Bitmap): Uri {
        val bytes = ByteArrayOutputStream()
        bitmap.compress(Bitmap.CompressFormat.JPEG, 100, bytes)
        val path = MediaStore.Images.Media.insertImage(contentResolver, bitmap, "Item_Photo", null)
        return Uri.parse(path)
    }

    private fun removePhoto() {
        selectedImageUri = null
        uploadedImageUrl = ""
        binding.imagePreview.visibility = View.GONE
        binding.imagePreview.setImageDrawable(null)
        binding.uploadPrompt.visibility = View.VISIBLE
        binding.buttonRemovePhoto.visibility = View.GONE
    }

    private fun postItem() {
        val name = binding.inputItemName.text.toString().trim()
        val location = binding.inputLocation.text.toString().trim()
        val description = binding.inputDescription.text.toString().trim()
        val status = if (binding.radioLost.isChecked) "lost" else "found"

        if (name.isEmpty() || location.isEmpty()) {
            Toast.makeText(this, "Item Name and Location are required.", Toast.LENGTH_LONG).show()
            return
        }

        val userId = FirebaseAuth.getInstance().currentUser?.uid ?: "anonymous"

        // Convert selected date to ISO format
        val isoFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        val createdDate = isoFormat.format(selectedDate.time)

        binding.buttonSaveItem.isEnabled = false
        binding.buttonSaveItem.text = "Posting..."

        lifecycleScope.launch {
            // Upload image if selected
            if (selectedImageUri != null) {
                uploadedImageUrl = uploadImageToFirebase(selectedImageUri!!) ?: ""
            }

            val newItem = Item(
                name = name,
                location = location,
                description = description,
                status = status,
                createdBy = userId,
                createdDate = createdDate,
                imageUrl = uploadedImageUrl
            )

            val result = ItemRepository.createItem(newItem)

            binding.buttonSaveItem.isEnabled = true
            binding.buttonSaveItem.text = "Submit"

            if (result.isSuccess) {
                Toast.makeText(this@AddEditItemActivity, "Item posted successfully!", Toast.LENGTH_SHORT).show()
                setResult(RESULT_OK)
                finish()
            } else {
                Toast.makeText(this@AddEditItemActivity, "Failed to post item: ${result.exceptionOrNull()?.message}", Toast.LENGTH_LONG).show()
            }
        }
    }

    private suspend fun uploadImageToFirebase(imageUri: Uri): String? {
        return try {
            val storageRef = FirebaseStorage.getInstance().reference
            val imageRef = storageRef.child("item_images/${UUID.randomUUID()}.jpg")

            imageRef.putFile(imageUri).await()
            val downloadUrl = imageRef.downloadUrl.await()
            downloadUrl.toString()
        } catch (e: Exception) {
            e.printStackTrace()
            null
        }
    }
}